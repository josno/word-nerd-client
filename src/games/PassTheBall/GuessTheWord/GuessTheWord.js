import React, { useState, useEffect } from "react";
import "./GuessTheWord.css";
import { Link } from "react-router-dom";
import GamesService from "../../../services/api-service";
import AnswerPage from "../AnswerPage/AnswerPage";
import ScrambledWordPage from "../ScrambledWordPage/ScrambledWordPage";
import PassTheBall from "../PassTheBall/PassTheBall";

const GuessTheWord = (props) => {
	const [showAnswer, setShowAnswer] = useState(false);
	const [passTheBall, setPassTheBall] = useState(true);
	const [randomWord, setRandomWord] = useState("");
	const [wordList, setWordList] = useState([]);

	const shuffleWord = (string) => {
		if (string.length <= 3) {
			const shortWordArray = string.split("");
			const shuffledShortWord =
				shortWordArray[1] + shortWordArray[0] + shortWordArray[2];
			return shuffledShortWord;
		} else {
			const unshuffled = string.split("");

			const shuffleString = function (str) {
				const shuffledWord = str
					.map((a) => ({ sort: Math.random(), value: a }))
					.sort((a, b) => a.sort - b.sort)
					.map((a) => a.value);
				if (shuffledWord === str) {
					shuffleString(shuffledWord);
				}
				return shuffledWord;
			};

			return shuffleString(unshuffled);
		}
	};

	const renderAnswer = () => {
		/* Don't show answer unless there's already a scrambled word*/
		if (passTheBall === false) {
			setShowAnswer(true);
		}
	};

	useEffect(() => {
		GamesService.getGameContent(props.gameId).then((responsejson) =>
			setWordList([...responsejson.word_list])
		);
	}, [props.gameId]);

	const handleClick = () => {
		const randomWord = wordList[Math.floor(Math.random() * wordList.length)];

		setRandomWord(randomWord);
		setPassTheBall(!passTheBall);
		showAnswer && setShowAnswer(!showAnswer);
	};

	const shuffledWord = shuffleWord(String(randomWord));

	return (
		<main className='guess-the-word-container'>
			{showAnswer ? (
				<AnswerPage word={randomWord} />
			) : passTheBall ? (
				<PassTheBall />
			) : (
				<ScrambledWordPage shuffledWord={shuffledWord} />
			)}

			<section className='game-controls'>
				<Link to={`/end-game-page`}>
					<button aria-label='end' className='control-button'>
						End
					</button>
				</Link>
				{!passTheBall && (
					<button
						aria-label='answer'
						className='control-button'
						onClick={() => renderAnswer()}
					>
						Answer
					</button>
				)}

				<button
					aria-label='next'
					className='control-button'
					onClick={() => handleClick()}
				>
					Next
				</button>
			</section>
		</main>
	);
};

export default GuessTheWord;
