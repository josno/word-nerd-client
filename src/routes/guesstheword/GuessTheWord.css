.game-controls {
	display: flex;
	justify-content: space-around;
	padding: 30px 0 30px 0;
	background-color: #18c0f0;
}

.game-play-container,
.scrambled-word-container,
.answer-page-container {
	margin: 10px;
}

.guess-the-word-container {
	height: 100vh;
	background-color: #18c0f0;
	padding-top: 30px;
}

.control-button {
	width: 20vw;
	height: 8vh;
	font-size: 1rem;
	border: 2px solid #d81159;
	border-radius: 15px;
	color: white;
	background-color: #0c0a3e;
	font-weight: bold;
}

.control-button:hover {
	cursor: pointer;
}

@media (min-width: 1000px) {
	.guess-the-word-container {
		padding-top: 100px;
	}
}

@media (min-width: 700px) {
	.game-play-container,
	.scrambled-word-container,
	.answer-page-container {
		height: 60vh;
		margin-top: 10vh;
	}
}
