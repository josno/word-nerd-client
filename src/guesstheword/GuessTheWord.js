import React, { Component } from 'react';
import WordContext from '../WordContext';
import './GuessTheWord.css';
import { Link } from 'react-router-dom';

class GuessTheWord extends Component {
	static contextType = WordContext;
	constructor(props) {
		super(props);
		this.state = {
			wordList: [],
			filteredList: []
		};
		this.shuffleWord = this.shuffleWord.bind(this);
	}

	shuffleWord(string) {
		const stringToArray = string.split('');
		const unshuffled = stringToArray;
		const shuffled = unshuffled
			.map(a => ({ sort: Math.random(), value: a }))
			.sort((a, b) => a.sort - b.sort)
			.map(a => a.value);
		return shuffled;
	}

	componentDidMount() {
		const { savedGames, currentGameId, filteredList } = this.context;
		const currentGame = savedGames.find(i => currentGameId === i.game_id);
		const currentWordList = currentGame.words;
		this.setState({
			wordList: [...currentWordList],
			filteredList: [...filteredList]
		});
	}

	render() {
		const { wordList, filteredList } = this.state;
		console.log(wordList, filteredList);

		const randomWord =
			wordList[Math.floor(Math.random() * wordList.length)];

		const shuffledWord = this.shuffleWord(String(randomWord));

		return (
			<div className="guess-the-word-container">
				<h1 className="pass-the-ball">Say The Word!</h1>
				<p className="scrambled-word">{shuffledWord}</p>

				<div className="game-controls">
					<Link to="/end-game-page">
						<button className="game-end">End Game</button>
					</Link>
					<Link to="/answer-page">
						<button
							className="answer-button"
							onClick={() =>
								this.context.getCurrentWord(randomWord)
							}
						>
							Answer
						</button>
					</Link>
					<Link to="/pass-the-ball">
						<button
							className="next-button"
							onClick={() =>
								this.context.makeFilteredList(
									randomWord,
									wordList
								)
							}
						>
							Next
						</button>
					</Link>
				</div>
			</div>
		);
	}
}

export default GuessTheWord;
