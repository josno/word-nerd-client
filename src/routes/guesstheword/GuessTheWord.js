import React, { Component } from 'react';
import './GuessTheWord.css';
import { Link } from 'react-router-dom';
import GamesService from '../../services/api-service';
import AnswerPage from '../../components/AnswerPage/AnswerPage';
import ScrambledWordPage from '../../components/ScrambledWordPage/ScrambledWordPage';
import PassTheBall from '../../components/PassTheBall/PassTheBall';

class GuessTheWord extends Component {
	constructor(props) {
		super(props);
		this.state = {
			showAnswer: false,
			passTheBall: true,
			randomWord: '',
			wordList: []
		};
		this.shuffleWord = this.shuffleWord.bind(this);
		this.renderAnswer = this.renderAnswer.bind(this);
		this.handleClick = this.handleClick.bind(this);
	}

	shuffleWord(string) {
		if (string.length <= 3) {
			const shortWordArray = string.split('');
			const shuffledShortWord =
				shortWordArray[1] + shortWordArray[0] + shortWordArray[2];
			return shuffledShortWord;
		} else {
			const stringToArray = string.split('');
			const unshuffled = stringToArray;
			const shuffled = unshuffled
				.map(a => ({ sort: Math.random(), value: a }))
				.sort((a, b) => a.sort - b.sort)
				.map(a => a.value);
			return shuffled;
		}
	}

	renderAnswer = () => {
		/* Don't show answer unless there's already a scrambled word*/
		if (this.state.passTheBall === false) {
			this.setState({
				showAnswer: true
			});
		}
	};

	componentDidMount() {
		GamesService.getGameContent(this.props.gameId)
			.then(responsejson =>
				this.setState({
					wordList: [...responsejson.word_list]
				})
			)
			.then(list => console.log('check'));
	}

	handleClick = () => {
		const { wordList } = this.state;
		const randomWord =
			wordList[Math.floor(Math.random() * wordList.length)];

		this.setState({
			randomWord: randomWord,
			showAnswer: false,
			passTheBall: !this.state.passTheBall
		});
	};

	render() {
		const { randomWord, showAnswer, passTheBall } = this.state;

		const shuffledWord = this.shuffleWord(String(randomWord));

		return (
			<div className="guess-the-word-container">
				{showAnswer ? (
					<AnswerPage word={randomWord} />
				) : passTheBall ? (
					<PassTheBall />
				) : (
					<ScrambledWordPage shuffledWord={shuffledWord} />
				)}

				<div className="game-controls">
					<Link to={`/end-game-page`}>
						<button className="control-button">End</button>
					</Link>
					{this.state.renderAnswer ? (
						''
					) : (
						<button
							className="control-button"
							onClick={this.renderAnswer}
						>
							Answer
						</button>
					)}

					<button
						className="control-button"
						onClick={this.handleClick}
					>
						Next
					</button>
				</div>
			</div>
		);
	}
}

export default GuessTheWord;
