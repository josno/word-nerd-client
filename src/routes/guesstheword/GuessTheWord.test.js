import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import GuessTheWord from './GuessTheWord';

describe('<GuessTheWord/>', () => {
	it('renders as expected', () => {
		const tree = renderer
			.create(
				<BrowserRouter>
					<GuessTheWord />
				</BrowserRouter>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
