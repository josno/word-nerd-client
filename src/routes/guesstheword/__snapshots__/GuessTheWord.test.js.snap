// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GuessTheWord/> renders as expected 1`] = `
<main
  className="guess-the-word-container"
>
  <section
    className="game-play-container"
  >
    <header
      className="scrambled-message"
    >
      <h1
        className="scrambled-message-text"
      >
        Start!
      </h1>
    </header>
    <div
      className="ball-image-container"
    >
      <img
        alt="ball-placeholder"
        className="ball-gif"
        src="giphy.gif"
      />
    </div>
  </section>
  <section
    className="game-controls"
  >
    <a
      href="/end-game-page"
      onClick={[Function]}
    >
      <button
        aria-label="end"
        className="control-button"
      >
        End
      </button>
    </a>
    <button
      aria-label="next"
      className="control-button"
      onClick={[Function]}
    >
      Next
    </button>
  </section>
</main>
`;
